// vue/src/platforms/weex/runtime/node-ops-compiled.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createElement = createElement;
exports.createElementNS = createElementNS;
exports.createTextNode = createTextNode;
exports.createComment = createComment;
exports.insertBefore = insertBefore;
exports.removeChild = removeChild;
exports.appendChild = appendChild;
exports.parentNode = parentNode;
exports.nextSibling = nextSibling;
exports.tagName = tagName;
exports.setTextContent = setTextContent;
exports.setAttribute = setAttribute;
/* globals renderer */
// renderer is injected by weex factory wrapper

var namespaceMap = exports.namespaceMap = {};

function createElement(tagName) {
  return new renderer.Element(tagName);
}

function createElementNS(namespace, tagName) {
  return new renderer.Element(namespace + ':' + tagName);
}

function createTextNode(text) {
  return new renderer.TextNode(text);
}

function createComment(text) {
  return new renderer.Comment(text);
}

function insertBefore(node, target, before) {
  if (target.nodeType === 3) {
    if (node.type === 'text') {
      node.setAttr('value', target.text);
      target.parentNode = node;
    } else {
      var text = createElement('text');
      text.setAttr('value', target.text);
      node.insertBefore(text, before);
    }
    return;
  }
  node.insertBefore(target, before);
}

function removeChild(node, child) {
  if (child.nodeType === 3) {
    node.setAttr('value', '');
    return;
  }
  node.removeChild(child);
}

function appendChild(node, child) {
  if (child.nodeType === 3) {
    if (node.type === 'text') {
      node.setAttr('value', child.text);
      child.parentNode = node;
    } else {
      var text = createElement('text');
      text.setAttr('value', child.text);
      node.appendChild(text);
    }
    return;
  }

  node.appendChild(child);
}

function parentNode(node) {
  return node.parentNode;
}

function nextSibling(node) {
  return node.nextSibling;
}

function tagName(node) {
  return node.type;
}

function setTextContent(node, text) {
  node.parentNode.setAttr('value', text);
}

function setAttribute(node, key, val) {
  node.setAttr(key, val);
}
