// less/lib/less/visitors/join-selector-visitor-compiled.js
"use strict";

var Visitor = require("./visitor");

var JoinSelectorVisitor = function JoinSelectorVisitor() {
    this.contexts = [[]];
    this._visitor = new Visitor(this);
};

JoinSelectorVisitor.prototype = {
    run: function run(root) {
        return this._visitor.visit(root);
    },
    visitRule: function visitRule(ruleNode, visitArgs) {
        visitArgs.visitDeeper = false;
    },
    visitMixinDefinition: function visitMixinDefinition(mixinDefinitionNode, visitArgs) {
        visitArgs.visitDeeper = false;
    },

    visitRuleset: function visitRuleset(rulesetNode, visitArgs) {
        var context = this.contexts[this.contexts.length - 1],
            paths = [],
            selectors;

        this.contexts.push(paths);

        if (!rulesetNode.root) {
            selectors = rulesetNode.selectors;
            if (selectors) {
                selectors = selectors.filter(function (selector) {
                    return selector.getIsOutput();
                });
                rulesetNode.selectors = selectors.length ? selectors : selectors = null;
                if (selectors) {
                    rulesetNode.joinSelectors(paths, context, selectors);
                }
            }
            if (!selectors) {
                rulesetNode.rules = null;
            }
            rulesetNode.paths = paths;
        }
    },
    visitRulesetOut: function visitRulesetOut(rulesetNode) {
        this.contexts.length = this.contexts.length - 1;
    },
    visitMedia: function visitMedia(mediaNode, visitArgs) {
        var context = this.contexts[this.contexts.length - 1];
        mediaNode.rules[0].root = context.length === 0 || context[0].multiMedia;
    },
    visitDirective: function visitDirective(directiveNode, visitArgs) {
        var context = this.contexts[this.contexts.length - 1];
        if (directiveNode.rules && directiveNode.rules.length) {
            directiveNode.rules[0].root = directiveNode.isRooted || context.length === 0 || null;
        }
    }
};

module.exports = JoinSelectorVisitor;
